### PR TITLE
Sm/zip reads readables earlier

### LINKED ISSUE #578: EMFILE: too many open files
Code Snippet
```
    const set = await ComponentSet.fromManifest({
        manifestPath: './package.xml',
        resolveSourcePaths: ['force-app']
    });

    // Start a deploy with the components
    const deploy = await set.deploy({ usernameOrConnection });
```

Failing Line

```
const deploy = await set.deploy({ usernameOrConnection });
```

Result

```
Error: EMFILE: too many open files, open 'force-app/main/default/objectTranslations/Account-de/Request_Code__c.fieldTranslation-meta.xml'
```

**Node Version**
v16.13.1

**OS**
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.3 LTS
Release:        20.04
Codename:       focal

**Package Version**

"@salesforce/source-deploy-retrieve": "^5.9.9"

**Suggestion**

Avoid using sync methods that graceful-fs does not overwrite like:

openSync
closeSync

### PR DESCRIPTION
**Pull request by mshanemc**
### What does this PR do?

zip operation is probably the most common thing SDR does (every push/deploy)
ZipWriter passed all the WriteInfos to Archiver as Readable
Before this PR, all those Readable remained open until the zip finalizes.
So that would hit windows file limits but only on the zip operations (source->metadata and metada->source formats worked fine).  

This PR moves the reading of those files **left** in the pipeline, so that they're read into buffers as they come into ZipWriter (using existing utility code) and then passed to Archiver.  Once a file is read into buffer, the stream closes and the file doesn't have to stay open.

Besides solving the bug for windows users, this also has some perf benefits for all users (compare the passing scale nut from the 2 circle jobs shown below, or the checked in perf stats for linux/mac) on the larger projects' `sourceToZip`

fix #578 . 

### What issues does this PR fix or reference?

[@W-11492994@](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000012HtqzYAC/view)
fix https://github.com/forcedotcom/cli/issues/1555 

### Functionality Before

https://app.circleci.com/pipelines/github/forcedotcom/source-deploy-retrieve/4345/workflows/e4e5f01c-5eaa-4b9c-becb-1d23dcf358a0/jobs/19752)

### Functionality After

https://app.circleci.com/pipelines/github/forcedotcom/source-deploy-retrieve/4351/workflows/f664546c-54fd-430e-a8ab-4a16dfec1d87/jobs/19817

### CODE REVIEW

**Review by WillieRuemmele — COMMENTED** [review]
couldn't get those tests working?

any idea why the `sourceToMdapi` time increased across the board? While the other three decreased?

**Comment by mshanemc** [comment]
> couldn't get those tests working?
> 
> any idea why the `sourceToMdapi` time increased across the board? While the other three decreased?

got them back now.  It's the typical UT scenario--the tests are deeply tied to the implementation.  I just wanted the NUTs to run to tell me I wasn't crazy (and they caught a real bug in folder types) before investing the time rewriting the UT

**Comment by mshanemc** [comment]
> any idea why the `sourceToMdapi` time increased across the board? While the other three decreased?

whatever it is, it's not affected by this change to the zip writer.  anything that's < 1 second of change I'm gonna consider noise (after observing these for a while).

**Review by RodEsp — CHANGES_REQUESTED** [review]
I was successfully able to run the perf NUTs with the new code and would intermittently get the "too many files open" error with the old code. Something to note though is that in both scenarios they took 30+ minutes to run 😕

I just have two very small comments but after those are addressed feel free to merge this at your convenience @mshanemc.

File: .circleci/config.yml [hunk 1]
```diff
@@ -177,13 +177,15 @@ workflows:
...
-            parameters:
-              os:
-                - linux
-                - windows
-          # required because this will commit to github
-          context: salesforce-cli
+          os: linux
+      # I wish we could *also* record the CI perf for windows, but I can't figure out the syntax
+      # until then, run as a separate job to test scaling for larger repos
+      - release-management/test-nut:
```
[the remark this answers is not in the document]
Comment (mshanemc) [reply .circleci/config.yml:184]: good idea.  I made a change to do that

**Pull request merged**